### ISSUE TITLE
How to use the ML model in real (live) environment.

### ISSUE THREAD

**Issue by Anurag0987** [0x00000000000000000000/NIDS-Using-UNSW-NB15#1]
Hi, I really appreciate your work on the paper "Enhanced Network Intrusion Detection System" and am curious to know how can we implement this work in a live environment.
I am trying to build an Anomaly-based network intrusion detection system and have gone through many research papers quite similar to yours, and I wanted to build a model capable of detecting any threats in the live network traffic with low false-positive rates. 
But I am stuck at the point of how can I get the input parameters to pass to the ML model for prediction. The features like proto, sports, dsport, TTL, win I am able to get these from the TCPDUMP tool, but other features like (ct_srv_src, ct_srv_dst, smeanz, sload, ct_dst_src_ltm, dmeanz, ct_src_ltm, synack, dload) are really hard to extract from live network traffic. Even the dataset has features like sttl, dttl, swin, dwin but I am only getting values like win and ttl (i.e., I am not able to get the source and destination differently). I have further used the Zeek IDS tool to get some extra features but ended up with features like sbytes, dbytes, state, service which I think will as less value than the above features to the output label.
 
I would really appreciate it if you can share your thoughts on the above issue and help me to build a real-time NIDS system.
Thank you.

**Comment by 0x00000000000000000000** [comment]
Hi, Anurag. Have you tried using Argus in conjunction with Zeek and TCPDump? Check the man page for Argus [here](https://www.systutorials.com/docs/linux/man/1-ra/) and scroll down to the flag -s. You will see a couple of features you can use there and match the required features needed with the dataset itself. Although normally it's a good idea to take the features that hold the maximum importance, I believe eliminating other features would generally do harm to FP and FN rates in this particular case in my experience. Since the features passed aren't too much (49) and if you use algorithms such as XGBoost that do not require one-hot encoding to be done or any normalization to be done (you can research this yourself and form your own opinions on this concept), the cost-effectiveness isn't affected too much either. I would suggest creating a spark pipeline that does this. Check [this](https://xgboost.readthedocs.io/en/stable/jvm/xgboost4j_spark_tutorial.html) for more information on how you may be able to do this. I would suggest you do all this development and testing in a fresh environment perhaps by using pyenv or simply a virtual machine. It would be lovely to see what you achieve using my work and other work on this to create a real-time implementation. Good luck to you. If you have further questions, you can reply on this very thread itself.

**Comment by 0x00000000000000000000** [comment]
Hi. I remember these features creating a problem in generating a validation set. While I was referring to the original work, I noticed features that were additionally generated since the used tools generated all the features they could. The work mentions using 12 C# Algorithms as well to generate features. Which basically meant that I had to generate these features myself. I don't have the code for this and the original dataset creator didn't provide any code for that either. 
For example, take "is_sm_ips_ports". The description given is "When the source and destination IP Address and the port are the same, the value is 1 else 0". So essentially you could generate this feature directly from the rows in your datasets. Features under the "ct" category are also relatively intuitive. Except these features would need to be created by referencing rows back and forth for creation. The idea is that these features were created with custom means so you can leverage Python or even Bash to create these. I hope that provides a direction at least. 

One idea for implementation could be to use NSL-KDD or some other dataset to make up for the missing values as well. I remember researching tools such as CICFlowmeter and Netflow that could help you with that in conjunction with whatever dataset you come across that seems good enough. I didn't do this but it would be interesting to see if that provides better accuracy.
